src/main/java/net/minecraftforge/common/ForgeVersion.java

```diff
@@ -204,6 +209,46 @@ public void run()
                 }
             }
 
+            /**
+             * Opens stream for given URL while following redirects (max. 5)
+             */
+            private InputStream openUrlStream(URL url) throws IOException
+            {
+                URL originalUrl = url;
+                int redirects = 0;
+                do
+                {
+                    URLConnection c = url.openConnection();
+                    boolean redirected = false;
+
+                    if (c instanceof HttpURLConnection)
+                    {
+                        HttpURLConnection huc = (HttpURLConnection) c;
+                        int responseCode = huc.getResponseCode();
+                        if (responseCode >= 301 && responseCode <= 303 || responseCode == 307 || responseCode == 308)
+                        {
+                            String loc = huc.getHeaderField("Location");
+                            url = new URL(url, loc);
+                            redirected = true;
+                            huc.disconnect();
+                        }
+                    }
+
+                    if (redirected)
+                    {
+                        if (++redirects >= MAX_HTTP_REDIRECTS)
+                        {
+                            throw new IOException("Too many redirects while trying to fetch " + originalUrl);
+                        }
+                    }
+                    else
+                    {
+                        return c.getInputStream();
+                    }
+                }
+                while (true);
```

Comment:
Can you write this without an infinite loop?
Maybe `for (int redirects = 0; redirects < MAX_HTTP_REDIRECTS; redirects++)` ?

Reply:
```java

Reply:
I'd suggest
```java
private InputStream openUrlStream(URL url) throws IOException
{
	URL contextURL = url;
	for (int redirects = 0; redirects < MAX_HTTP_REDIRECTS; redirects++)
	{
		URLConnection c = contextURL.openConnection();
		if (c instanceof HttpURLConnection)
		{
			HttpURLConnection huc = (HttpURLConnection) c;
			huc.setInstanceFollowRedirects(false);
			int response = huc.getResponseCode();
			if (response >= 300 && response <= 399)
			{
				String loc = huc.getHeaderField("Location");
				try
				{
					contextURL = new URL(contextURL, loc);
				}
				finally
				{
					huc.disconnect();
				}
				continue;
			}
		}
		return c.getInputStream();
	}
	throw new IOException("Too many redirects while trying to fetch: " + url);
}
```